src/main/java/pl/javastart/couponcalc/PriceCalculator.java

```diff
@@ -1,13 +1,75 @@
 package pl.javastart.couponcalc;
 
+import java.math.BigDecimal;
+import java.math.RoundingMode;
 import java.util.List;
 
 public class PriceCalculator {
 
     public double calculatePrice(List<Product> products, List<Coupon> coupons) {
-        // TODO
+        double sum = 0;
+        if (coupons == null && products == null) {
+            return 0.0;
+        }
+        if (coupons == null) {
+            return getSumPriceWithoutDisc(products);
+        }
+        if (coupons.size() == 1) {
+            Coupon coupon = coupons.get(0);
+            if (!checkIfCouponHasCategory(coupon)) {
+                return getSumPriceWithCouponForAll(products, coupon);
+            } else {
+                return getSumPriceWithCouponWithCat(products, coupon);
+            }
+        } else {
+            double sumWithoutDisc = getSumPriceWithoutDisc(products);
+            double theBestDiff = 0;
+            for (Coupon coupon : coupons) {
+                double sumWithDisc = getSumPriceWithCouponWithCat(products, coupon);
+                double tempDiff = sumWithoutDisc - sumWithDisc;
+                if (tempDiff > theBestDiff) {
+                    theBestDiff = tempDiff;
+                }
+            }
+            sum = sumWithoutDisc - theBestDiff;
+        }
```

Comment:
Niepotrzebnie rozróżniasz sytuację dla 1 lub wielu kuponów. W obu przypadkach algorytm powinien być dokładnie taki sam:
1. Weź kolejny kupon
2. oblicz dla niego sumę cen
3. jeżeli suma jest najmniejsza to ją zapamiętaj
4. powtarzaj 1-3 dla wszystkich kuponów

Czyli niezależnie od tego, czy będzie 1, czy wiele kuponów, to pętla powinna się wykonać 1, albo wiele razy.